sopel/plugin.py

```diff
@@ -946,6 +956,54 @@ def actual_decorator(function):
     return actual_decorator
 
 
+def url_lazy(loader):
+    """Decorate a function to handle URL, using lazy-loading for its regex.
+
+    :param loader: a callable to generate a list of regexes when the bot is
+                   loading this URL callback
+    :type loader: :term:`function`
+
+    The ``loader`` function must accept a ``settings`` parameter and return a
+    list (or tuple) of compiled regular expressions::
+
+        import re
+
+        def loader(settings):
+            return [re.compile(r'<your_url_pattern>')]
+
+    It will be called by Sopel when the bot parses the plugin to register URL
+    callbacks to get its regexes. The ``settings`` argument will be the bot's
+    :class:`sopel.config.Config` object.
+
+    If the ``loader`` function raises an
+    :exc:`~sopel.plugins.exceptions.ImproperlyConfigured` exception, the URL
+    callback will be ignored. This exception can be used with a message that
+    will be logged as an error; it will not fail the plugin's loading.
+
+    The decorated function will behave like any other :func:`callable`::
+
+        from sopel import plugin
+
+        @plugin.url_lazy(loader)
+        def my_url_handler(bot, trigger):
+            bot.say('URL found: %s' % trigger.group(0))
+
+    .. versionadded:: 7.1
+
+    .. important::
+
+        There are few differences with the :func:`url` decorator:
+
+        * the decorated callable does **not** have a ``match`` parameter
+        * the decorated callable accepts **one and only one** ``loader``
```

Comment:
I doubt it'd be too difficult to accept multiple `loader`s for parity with the other decorators.

We still have #1619 open for 8.0, and let's not make more work for that one. 😁

Reply:
I don't want to, for 2 reasons:

* as a function, a dev can compose it however they want: calling another function, generating a specific list, etc. whatever they fancy
* I don't want to deal with corner cases, like "oh, one of the 3 loaders raised an exception, do we ignore it and keep the other? do we crash even if only one failed?" or something like "which one of the 10 loaders failed to deliver?" or "if there is the same pattern twice from 2 different loaders, what do we do?" or "if there are two patterns that are different but match the same thing, how should the decorator handle them?"

Lazy loading is a powerful feature, and as such, can be quite dangerous. I prefer to limit it in a way that the dev is ultimately both in control of and responsible for the outcome.

Reply:
About corner cases, doesn't the same potential exist for e.g. one of several `rule` patterns to be invalid regex and fail to compile? Actually, all of those issues in the second point have counterparts in the existing multi-use/multi-param decorators (except for the identical-pattern one, maybe). Even a single loader function could generate multiple patterns that match the same thing by accident (or design), so why is that considered a corner case? Why not just copy the behavior of the other decorators—if something doesn't load properly, the plugin doesn't register. You're done, no need to handle corner cases.

I see where you're coming from on this, but it would be weird for the plugin API to explicitly prevent shooting yourself in the foot _only here_ when it's allowed everywhere else. Helping the user (plugin dev) avoid shooting their own foot sounds appealing, but there's only so much safety possible without somehow embedding an AGI into Sopel's rule manager. 😁

Reply:
Indeed, we can't prevent all the bullet-to-foot possibilities. After all, we are humans, and so are the users. So for me, it's about the trade off we can offer: how far can you go with one bullet? One foot? Both feet? OK, enough of this analogy. 😛 

I consider that the trade-off is as follow:

* option 1 (`url`): you get as many patterns as you want, but they are static (not limit but less power). I like it because you can statically check that the patterns are the right ones.
* option 2 (`url_lazy`): you can define the regexes when the software is running, but you can't define more than one loader (more power but strict limit). I like it because you have to test one and only one function. You don't have to run cross-validation tests or anything. You still have to make sure that all the possible states of your function are accounted for, but at least you have only one function to test.

I guess I think with unit tests in mind first, and I try to get the appropriate trade-off between having power and being error prone.

Reply:
This is just to explain where I'm coming from. I threw it together, and it's not necessarily the _most_ complete argument, but I wanted to make sure this doesn't get forgotten (since it's been 4 days since we last chatted about it!)

On the other hand, I can see a plugin developer wanting to reuse certain pieces of regex-generating functionality. It's not even a toy example, but the current implementation (only one loader, period) forces:

```py
def _lazy_spam(settings):
    # figure out spam rules

def _lazy_eggs(settings):
    # figure out eggs rules

def _loader_func_eggspam(settings):
    for pattern in itertools.chain(_lazy_spam(settings), _lazy_eggs(settings)):
        yield pattern

@plugin.url_lazy(_lazy_spam)
def spam_only(bot, trigger):
    # whatever this does

@plugin.url_lazy(_loader_func_eggspam)
def spam_and_eggs(bot, trigger):
    # whatever happens here
```

With multiple loaders/decorators allowed for `url_lazy`, it's certainly cleaner to read:

```py
def _lazy_spam(settings):
    # figure out spam rules

def _lazy_eggs(settings):
    # figure out eggs rules

@plugin.url_lazy(_lazy_spam)
def spam_only(bot, trigger):
    # whatever this does

@plugin.url_lazy(_lazy_spam, _lazy_eggs)
def spam_and_eggs(bot, trigger):
    # whatever happens here

# Alternatively, we might want to leave additional positional args to `url_lazy`
# available for future enhancements, like passing extra args along to the loader
@plugin.url_lazy(_lazy_spam)
@plugin.url_lazy(_lazy_eggs)
def spam_and_eggs(bot, trigger):
    # same thing
```

On the _other_ other hand, I can see where implementing and documenting it as "one and only one loader function" _now_ doesn't preclude a later change to allow multiple—but allowing multiple from the start ties our hands later due to back-compat goals. I'm not saying that multi-loader support is a blocker for accepting this patch or anything, but I'm sure if I actually sat down and thought for more than a few minutes I could devise a real use-case for sharing loader logic like this. Whether a first-class function (like I did) or a lambda, it's pretty ugly either way if that's the only means of doing it.

Reply:
Maybe all we need is more tools then? Have a look at this one:

https://github.com/sopel-irc/sopel/pull/1904/commits/e55eb1c4766a5bb6a49b4dc14a2bd0b47ad9eb37

Reply:
Note that the same logic can be used with variations:

* make a chain loaders that check for uniqueness
* have a `rule_lazy` decorator that will behave like `url_lazy`... and the chain loaders function can be reused *without modification*!

Reply:
Hmm, this could be a good approach in theory. I'd still prefer if such `_lazy` decorators just accepted multiple arguments directly, though.

Also: This `chain_loaders` function belongs somewhere else, probably in `tools`. `plugin` should contain decorators only.

Reply:
Yes `tools` sound like a better idea indeed!

I don't think it's good only in theory, I find it more practical to tell developers, by design, that they need to be mindful about how to merge the result of more than one loader. It also begs the question why someone would need to merge loaders, that are probably already used by another rule(s).

It feels like an edge case: who would like to have 2 rules that trigger on the same line, in the same plugin, for the same regex? Take the bugzilla, it's a good example for URL callbacks: you only need one, because the variation is about the domain name. For URLs, I believe this is probably the main if not the only reason to have a lazy loader. Domain name, base path variation, etc. that's probably it.

A use case I can think of is something like that:

```
def _regex_base(settings):
    return re.escape(settings.gitlab.my_gitlab_domain)

def _loader_pulls(settings):
    base = _regex_base(settings)
    return [base + pull_pattern]

def _loader_issue(settings):
    base = _regex_base(settings)
    return [base + issue_pattern]

@plugin.url_lazy(_loader_egg)
def pull_url(bot, trigger):
    # something something pull request

@plugin.url_lazy(_loader_spam)
def issue_url(bot, trigger):
    # something something issue
```

In that case, URL callback shares the domain name, which is known from a configuration value, but everything else depend on the type of URL callback you need.

I understand the other decorators accepts more than one pattern/name at a time, but I don't think they can be compared with their lazy versions.

Reply:
After further consideration and many conversations (with many people, including - yes - myself), I decided to follow @dgw's opinion. My conclusion to all of these conversations was: it's an edge case, if someone complain, there is nothing they can't fix with some more code.